// spsave/lib/src/utils/ConsoleLogger.js
"use strict";
var colors = require('colors');
var util = require('util');
var ConsoleLogger = (function () {
    function ConsoleLogger() {
    }
    ConsoleLogger.prototype.error = function (message) {
        this.log(message, 'red');
    };
    ConsoleLogger.prototype.warning = function (message) {
        this.log(message, 'yellow');
    };
    ConsoleLogger.prototype.success = function (message) {
        this.log(message, 'green');
    };
    ConsoleLogger.prototype.info = function (message) {
        this.log(message, 'white');
    };
    ConsoleLogger.prototype.log = function (message, color) {
        if (message === '') {
            console.log('');
            return;
        }
        var dateNow = new Date();
        var dateString = util.format('[%s:%s:%s]', ('0' + dateNow.getHours()).slice(-2), ('0' + dateNow.getMinutes()).slice(-2), ('0' + dateNow.getSeconds()).slice(-2));
        console.log(colors[color](dateString + " spsave: " + message));
    };
    return ConsoleLogger;
}());
exports.ConsoleLogger = ConsoleLogger;
